ls1intum/Artemis
# [Enhancement] Improve feedback and grading generation

### Checklist
- [x] Server: I updated multiple integration tests (Spring) related to the features (with a high test coverage)

### Motivation and Context



Preparing the grading and feedback system of programming exercises to be extended with the new static code analysis configuration of #2086.

### Description


- Cleanup and refactor feedback generation in bamboo and jenkins service.
    - Moved generation of test-case-feedback from BambooService to FeedbackService (now in-line with sca feedback generation)
    - Remove retrieveLatestResult method used for the 'trigger-failed-build' endpoint
        - the retrieved result had different information than the result sent to Artemis via the updated bamboo notification plugin
        - therefore the score generation would be false
        - the endpoint now **always** triggers a new build
- Add ProgrammingExerciseGradingService and ProgrammingExerciseGradingResource to handle grading specific api requests and score generation
    - *Most of the code is moved and adapted from the TestCaseService / -Resource classes.*
    - Moved **processNewProgrammingExerciseResult** from ResultService to GradingService
    - Moved **updateResult** and **updateAllResults** from TestCaseService to GradingService
    - Moved **reset** and **re-evaluate** endpoints from TestCaseResource to GradingResource
- **We do not yet include static code analysis in the score generation**.

### Steps for Testing


Check that the grading system for programming exercises still works as before.
Since this affects CI-specific code, please test this for both setups - Bamboo and Jenkins (TS1 / TS2).

1. Log in to Artemis
2. Create a Programming Exericse
3. Adjust the TestCase Configuration via the Grading page.
4. Participate in the exercise
5. Check that the score is calculated correctly and all test case feedback is displayed correctly.

Also test:
- Re-evaluating the results via the grading page.
- Triggering all builds via the grading page.
- Resetting the grading config via the grading page.
- Triggering a failed build as a student (Stop the build on bamboo/jenkins and refresh the page to see the red button).
- Tests after due date
- Sequential tests
- Test cases with bonus Points
- Participation after due date should still be possible with an unrated result
- Manual results (inline feedback feature)

### Test Coverage


- FeedbackService.java: 95%
- ProgrammingExerciseGradingService.java: 94%
- ProgrammingExerciseTestCaseService.java: 99%
- ProgrammingExerciseGradingResource.java: 97%
- ProgrammingExerciseTestCaseResource.java: 94%
